libpam/include/security/pam_sandbox.h

```diff
@@ -0,0 +1,136 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/landlock.h>
+#include <linux/prctl.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/prctl.h>
+#include <sys/stat.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#ifndef landlock_create_ruleset
+
+	static inline int
+landlock_create_ruleset(const struct landlock_ruleset_attr *const attr,
+		const size_t size, const __u32 flags)
+{
+	return syscall(__NR_landlock_create_ruleset, attr, size, flags);
+}
+#endif
+
+#ifndef landlock_add_rule
+static inline int landlock_add_rule(const int ruleset_fd,
+		const enum landlock_rule_type rule_type,
+		const void *const rule_attr,
+		const __u32 flags)
+{
+	return syscall(__NR_landlock_add_rule, ruleset_fd, rule_type, rule_attr,
+			flags);
+}
+#endif
+
+#ifndef landlock_restrict_self
+static inline int landlock_restrict_self(const int ruleset_fd,
+		const __u32 flags)
+{
+	return syscall(__NR_landlock_restrict_self, ruleset_fd, flags);
+}
+#endif
+
+#define ACCESS_FS_ROUGHLY_READ ( \
+		LANDLOCK_ACCESS_FS_EXECUTE | \
+		LANDLOCK_ACCESS_FS_READ_FILE | \
+		LANDLOCK_ACCESS_FS_READ_DIR)
+
+#ifdef LANDLOCK_ACCESS_FS_REFER
+#define ACCESS_FS_ROUGHLY_WRITE ( \
+		LANDLOCK_ACCESS_FS_WRITE_FILE | \
+		LANDLOCK_ACCESS_FS_REMOVE_DIR | \
+		LANDLOCK_ACCESS_FS_REMOVE_FILE | \
+		LANDLOCK_ACCESS_FS_MAKE_CHAR | \
+		LANDLOCK_ACCESS_FS_MAKE_DIR | \
+		LANDLOCK_ACCESS_FS_MAKE_REG | \
+		LANDLOCK_ACCESS_FS_MAKE_SOCK | \
+		LANDLOCK_ACCESS_FS_MAKE_FIFO | \
+		LANDLOCK_ACCESS_FS_MAKE_BLOCK | \
+		LANDLOCK_ACCESS_FS_MAKE_SYM | \
+		LANDLOCK_ACCESS_FS_REFER)
+
+#define ACCESS_ABI_2 ( \
+		LANDLOCK_ACCESS_FS_REFER)
+#else
+#define ACCESS_FS_ROUGHLY_WRITE ( \
+		LANDLOCK_ACCESS_FS_WRITE_FILE | \
+		LANDLOCK_ACCESS_FS_REMOVE_DIR | \
+		LANDLOCK_ACCESS_FS_REMOVE_FILE | \
+		LANDLOCK_ACCESS_FS_MAKE_CHAR | \
+		LANDLOCK_ACCESS_FS_MAKE_DIR | \
+		LANDLOCK_ACCESS_FS_MAKE_REG | \
+		LANDLOCK_ACCESS_FS_MAKE_SOCK | \
+		LANDLOCK_ACCESS_FS_MAKE_FIFO | \
+		LANDLOCK_ACCESS_FS_MAKE_BLOCK | \
+		LANDLOCK_ACCESS_FS_MAKE_SYM)
+#endif
+
+#define ACCESS_FILE_RW ( \
+		LANDLOCK_ACCESS_FS_WRITE_FILE | \
+		LANDLOCK_ACCESS_FS_READ_FILE)
+
+#define ACCESS_FILE_RO (LANDLOCK_ACCESS_FS_READ_FILE)
+
+
+static int _restrict_to_file_read(const char * const file_path)
```

Comment:
This `_restrict_to_file_read()` doesn't look like a good design because I think it will lead to calling it multiple times, which will create extra nested sandboxes, which are limited to 16 layers for the whole system. Only one layer is needed here, and the API should follow the builder pattern (as do the Landlock syscalls), i.e. create a ruleset, add some rules, and finally enforce them.

Reply:
Right, will split to match the builder pattern.

Reply:
Done